File: app/src/main/java/com/bopr/android/smailer/GeoCoordinates.java
package com.bopr.android.smailer;

import android.location.Location;
import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;
import androidx.core.util.ObjectsCompat;

/**
 * Geolocation coordinates.
 *
 * @author Boris Pronin (<a href="mailto:boprsoft.dev@gmail.com">boprsoft.dev@gmail.com</a>)
 */
public class GeoCoordinates implements Parcelable {

    private double latitude;
    private double longitude;

    public GeoCoordinates() {
        /* Default constructor required by Jackson */
    }

    public GeoCoordinates(double latitude, double longitude) {
        this.latitude = latitude;
        this.longitude = longitude;
    }

    public GeoCoordinates(Location location) {
        this(location.getLatitude(), location.getLongitude());
    }

    public double getLatitude() {
        return latitude;
    }

    public double getLongitude() {
        return longitude;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o)
            return true;
        if (o == null || getClass() != o.getClass())
            return false;

        GeoCoordinates that = (GeoCoordinates) o;
        return Double.compare(that.latitude, latitude) == 0 && Double.compare(that.longitude, longitude) == 0;
    }

    @Override
    public int hashCode() {
        return ObjectsCompat.hash(latitude, longitude);
    }

    @Override
    @NonNull
    public String toString() {
        return "GeoCoordinates{" +
                "latitude=" + latitude +
                ", longitude=" + longitude +
                '}';
    }

    /* Generated Parcelable stuff. Alt+Enter on "implements Parcelable" to update */

    protected GeoCoordinates(Parcel in) {
        latitude = in.readDouble();
        longitude = in.readDouble();
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeDouble(latitude);
        dest.writeDouble(longitude);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public static final Creator<GeoCoordinates> CREATOR = new Creator<GeoCoordinates>() {
        @Override
        public GeoCoordinates createFromParcel(Parcel in) {
            return new GeoCoordinates(in);
        }

        @Override
        public GeoCoordinates[] newArray(int size) {
            return new GeoCoordinates[size];
        }
    };


}

File: app/src/main/kotlin/com/bopr/android/smailer/GeoCoordinates.kt
package com.bopr.android.smailer

import android.location.Location
import android.os.Parcelable
import com.bopr.android.smailer.util.decimalToDMS
import kotlinx.parcelize.Parcelize

/**
 * Geolocation coordinates.
 *
 * @author Boris Pronin ([boprsoft.dev@gmail.com](mailto:boprsoft.dev@gmail.com))
 */
@Parcelize
data class GeoCoordinates(
        val latitude: Double = 0.0,
        val longitude: Double = 0.0) : Parcelable {

    fun format(degreeSymbol: String = "°",
               minuteSymbol: String = "\'",
               secondSymbol: String = "\"",
               northSymbol: String = "N",
               southSymbol: String = "S",
               westSymbol: String = "W",
               eastSymbol: String = "E",
               separator: String = ", "): String {
        return StringBuilder()
                .append(decimalToDMS(latitude, degreeSymbol, minuteSymbol, secondSymbol))
                .append(if (latitude > 0) northSymbol else southSymbol)
                .append(separator)
                .append(decimalToDMS(longitude, degreeSymbol, minuteSymbol, secondSymbol))
                .append(if (longitude > 0) westSymbol else eastSymbol)
                .toString()
    }

    companion object {

        fun coordinatesOf(location: Location?): GeoCoordinates? {
            return location?.run {
                coordinatesOf(latitude, longitude)
            }
        }

        fun coordinatesOf(latitude: Double?, longitude: Double?): GeoCoordinates? {
            return if (latitude != null && longitude != null) {
                GeoCoordinates(latitude, longitude)
            } else {
                null
            }
        }
    }
}
